ui/apps/platform/src/hooks/useWidgetConfig.test.tsx

```diff
@@ -0,0 +1,242 @@
+/* eslint-disable @typescript-eslint/no-floating-promises */
+import React from 'react';
+import { render, act as reactAct } from '@testing-library/react';
+import { renderHook, act as hooksAct } from '@testing-library/react-hooks';
+import useWidgetConfig, { defaultStorageKey } from 'hooks/useWidgetConfig';
+
+beforeEach(() => {
+    localStorage.clear();
+});
+
+describe('useWidgetConfig hook', () => {
+    it('should persist a widget configuration in localStorage', () => {
+        const widgetId = 'single-component';
+        const routeId = '/';
+        const defaultConfig = {
+            sort: 'asc',
+            filter: '',
+        };
+        const { result: firstResult } = renderHook(() =>
+            useWidgetConfig(widgetId, routeId, defaultConfig)
+        );
+
+        // Expect the initial state to be clean and return the provided defaults
+        expect(firstResult.current[0]).toEqual(defaultConfig);
+
+        // By default changes should merge the new and old configs
+        hooksAct(() => {
+            const [, updateConfig] = firstResult.current;
+            updateConfig({ filter: 'Namespace:stackrox' });
+        });
+        expect(firstResult.current[0]).toEqual({ sort: 'asc', filter: 'Namespace:stackrox' });
+
+        hooksAct(() => {
+            const [, updateConfig] = firstResult.current;
+            updateConfig({ sort: 'desc' });
+        });
+        expect(firstResult.current[0]).toEqual({ sort: 'desc', filter: 'Namespace:stackrox' });
+
+        // Loading the same widget/route config in a new component, i.e. a page reload
+        const { result: secondResult } = renderHook(() =>
+            useWidgetConfig(widgetId, routeId, defaultConfig)
+        );
+
+        // Results should be the persisted config instead of the default
+        expect(secondResult.current[0]).toEqual({ sort: 'desc', filter: 'Namespace:stackrox' });
+    });
+
+    it('should allow a custom update method to be used when updating the configuration', () => {
+        const widgetId = 'single-component';
+        const routeId = '/';
+        const defaultConfig = {
+            sort: 'asc',
+            filter: '',
+        };
+        type Config = typeof defaultConfig;
+
+        // Reducer that does a full config replace instead of merge
+        const reducer = (_oldConfig, payload: Partial<Config>) => payload;
+
+        const { result: firstResult } = renderHook(() =>
+            useWidgetConfig(widgetId, routeId, defaultConfig, reducer)
+        );
+
+        expect(firstResult.current[0]).toEqual(defaultConfig);
+
+        hooksAct(() => {
+            const [, updateConfig] = firstResult.current;
+            updateConfig({ filter: 'Namespace:stackrox' });
+        });
+
+        // Even though the `sort` property was overwritten, the hook ensures type safety by
```

Comment:
At first, I was thinking what if we want to clear the sort, but I guess technically that's something you should have regardless

Reply:
That's a good point, I think in that case I'd probably update the union type to contain something like "NO_SORT". The idea here was that since it is stored in localStorage we can't really be sure what we are getting back, so this does the best it can to ensure we don't run into weird runtime errors.